src/DSManager.cpp

```diff
@@ -405,15 +419,163 @@ Subscriber* DSManager::removeSubscriber(
     return ret;
 }
 
+ReturnCode_t DSManager::deleteSubscriber(
+        DataReader* dr)
+{
+    fastdds::dds::Subscriber* sub = nullptr;
+    {
+        std::lock_guard<std::recursive_mutex> lock(management_mutex);
+        GUID_t participant_guid = dr->get_subscriber()->get_participant()->guid();
+        sub = entity_map[participant_guid].subscriber;
+    }
+    return sub->delete_datareader(dr);
+
+}
+
+ReturnCode_t DSManager::deletePublisher(
+        DataWriter* dw)
+{
+    fastdds::dds::Publisher* pub = nullptr;
+    {
+        std::lock_guard<std::recursive_mutex> lock(management_mutex);
+        GUID_t participant_guid = dw->get_publisher()->get_participant()->guid();
+        pub = entity_map[participant_guid].publisher;
+    }
+    return pub->delete_datawriter(dw);
+
+}
+
+ReturnCode_t DSManager::deleteParticipant(
+        DomainParticipant* participant)
+{
+    {
+
+        std::lock_guard<std::recursive_mutex> lock(management_mutex);
+
+        auto& entity = entity_map[participant->guid()];
+
+        std::vector<DataWriter*> writers;
+        std::vector<DataReader*> readers;
+
+        entity.publisher->get_datawriters(writers);
+        entity.subscriber->get_datareaders(readers);
+
+        // First we clear the DataReaders/DataWriters
+        ReturnCode_t ret;
+
+        for (const auto& writer: writers)
+        {
+            ret = entity.publisher->delete_datawriter(writer);
+            if (ReturnCode_t::RETCODE_OK != ret)
+            {
+                LOG_ERROR("Error deleting Data Writer");
+            }
+            auto topic = writer->get_topic();
```

Comment:
After calling `delete_datawriter` I do not think it is safe to use the DataWriter pointer anymore. If you need this information, obtain it before calling `delete_datawriter`. However, type is the only information it is not essentially required to be unregistered in order to delete the domainParticipant and in most cases it is not usually done.

This also applies when deleting and unregistering the DataReader types.